### DISCUSSION TITLE
Temporary Tables and Sessions?

### DISCUSSION THREAD

**Discussion by SebastianSchroeder** [sqlalchemy/sqlalchemy#6905] [Usage Questions]
Hi there,

I am using sqlAlchemy 1.3.23 with psycopg2 to connect to redshift. As suggested in the documentation I am using a global engine for all connections to the DB. I am running into a problem with temporary tables when I do share the engine though.

My test code to reproduce the error is as follows:

```python
def test_temporary_tables(self):
  with session_scope(self.ctx) as session:
    session.execute("""CREATE TABLE #test AS SELECT * FROM table_2""")

  with session_scope(self.ctx) as session:
    session.execute("""CREATE TABLE #test AS SELECT * FROM table_2""")
```

The session_scope provides the context manager around the session. The error I get is the following:

```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.DuplicateTable) relation "#test" already exists

[SQL: CREATE TABLE #test AS SELECT * FROM table_2]
```

The temporary table is visible between sqlAlchemy sessions. I was able to fix the error by using the _NullPool_ as the poolclass for the engine, since then every new session gets a new connection as thus a new db session. I would like to reuse the connections though. 

I don't see a way of starting a new db session on every new sqlAlchemy Session while reusing the connection. Is there a way to do this?

**Comment by CaselIT** [comment]
Hi,

Could you provide the source of `session_scope`?

**Reply by SebastianSchroeder** [reply]
Sure, here it is:

```python

_redshift_engine = None
_redshift_session_maker = sessionmaker()

def redshift_engine(ctx):
    global _redshift_engine
    if _redshift_engine is None:
        conn_string = mk_conn_string(ctx.rs_database_user, ctx.rs_database_pwd, ctx.rs_host, 5439, ctx.rs_database)
        _redshift_engine = create_engine(
            conn_string,
            connect_args={'sslmode': 'require'},
            echo=is_debug()
        )
    return _redshift_engine


def redshift_session(ctx):
    if _redshift_session_maker.kw['bind'] is None:
        _redshift_session_maker.configure(bind=(redshift_engine(ctx)))
    return _redshift_session_maker

@contextmanager
def session_scope(ctx: ConfigurationContext) -> Session:
    session_maker = redshift_session(ctx)
    # create a Session
    session = session_maker()
    try:
        yield session
        session.commit()
    except:
        session.rollback()
        raise
    finally:
        session.close()

```

**Reply by CaselIT** [reply]
I get the same behaviour with postgress. At least on postgress it seems expected, see https://www.postgresql.org/docs/current/sql-createtable.html. A "session" in the documentation seems to be duration of the connection, not a transaction. You can specify `ON COMMIT DROP` on pg to make what you want (not sure about redshift, but it should be similar)

**Reply by SebastianSchroeder** [reply]
Thank you for validating the behaviour. Sadly it does not seem that redshift support it
https://docs.aws.amazon.com/redshift/latest/dg/r_CREATE_TABLE_NEW.html

I'll try a 'drop table if exists' at the start of the session or otherwise the NullPull will have to suffice.

**Comment by gordthompson** [comment]
Does `CREATE TABLE #test …` really create a temporary table in PostgreSQL? I thought that was just a SQL_Server/Sybase thing, and PostgreSQL used `CREATE TEMPORARY TABLE test …`

**Reply by CaselIT** [reply]
yes, you need to use temporary or temp after create. Redshift seems to extend it to accept a # prefix for temporary tables

**Comment by sebtrack** [comment]
I have the same issue, besides `NullPool` no other options?

**Reply by zzzeek** [reply]
i have  a suggestion, see answer below momentarily

**Comment by zzzeek** [comment]
hi  -

shouldn't they use [DISCARD](https://www.postgresql.org/docs/current/sql-discard.html) when connections are returned to the pool in order to make sure temp tables are removed?   This can be done as:

```py
@event.listens_for(SomeEngineOrPool, 'checkin')
def receive_checkin(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("DISCARD TEMPORARY")
    cursor.close()
```

**Reply by sebtrack** [reply]
Hm, this is an interesting solution, but it's not working for me. I debugged it, and the event is firing, the statement is executed on the cursor, yet the temporary tables I create live on. I might have to use unique names for the temp tables and clean up manually.

**Reply by sebtrack** [reply]
1. Request:

```
CREATE TEMPORARY TABLE temp_table AS ...
```

before `cursor.execute("DISCARD TEMPORARY")`

```
cursor.execute("SELECT count(*) FROM temp_table;")
cursor.fetchall()
```

returns `[(30079,)]`. Stepping to the next line `cursor.close()` and running the same as above, returns:

```
{UndefinedTable}relation "temp_table" does not exist
LINE 1: SELECT count(*) FROM temp_table;
                             ^
```

Voilà! But in the second web request, going through the same method:

2. Request:
```
app            | (psycopg2.errors.DuplicateTable) relation "temp_table" already exists
app            | 
app            | [SQL: CREATE TEMPORARY TABLE temp_table AS ...
app            | FROM table))]
app            | (Background on this error at: http://sqlalche.me/e/13/f405)
```

It's a Schroedingers table.

**Reply by zzzeek** [reply]
maybe call connection.commit()?

**Reply by sebtrack** [reply]
I use this Session/Context Manager:

```python
@contextmanager
def AutocommitScopedTsdbSessionContextManager(do_commit=True):
    with tsdb.engine.connect().execution_options(isolation_level='AUTOCOMMIT') as conn:
        Session = scoped_session(sessionmaker(bind=conn))
        session = Session()
        try:
            yield session
            if do_commit:
                session.commit()
        except:
            session.rollback()
            raise
        finally:
            session.close()
```

**Reply by zzzeek** [reply]
in the event hook, call commit

```py
@event.listens_for(SomeEngineOrPool, 'checkin')
def receive_checkin(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("DISCARD TEMPORARY")
    cursor.close()
    dbapi_connection.commit()
```

**Reply by sebtrack** [reply]
This is working! Wow, seems so easy and it still can become quite complex!
